tensorwaves/interfaces.py

```diff
@@ -0,0 +1,111 @@
+"""Defines top-level interfaces of tensorwaves."""
+
+from abc import ABC, abstractmethod
+
+
+class Function(ABC):
+    """
+    Interface of a callable function.
+
+    The parameters of the model are separated from the domain variables. This
+    follows the mathematical definition, in which a function defines its domain
+    and parameters. However specific points in the domain are not relevant.
+    Hence while the domain variables are the argument of the evaluation
+    (see :func:`~Function.__call__`), the parameters are controlled via a
+    getter and setter (see :func:`~Function.parameters`). The reason for this
+    separation is to facilitate the events when parameters have changed.
+    """
+
+    @abstractmethod
+    def __call__(self, dataset: dict) -> list:
+        """Evaluate the function.
+
+        Args:
+            dataset: a `dict` with domain variable names as keys.
+
+        Return:
+            `list` or `numpy.array` of values.
+        """
+
+    @property
+    @abstractmethod
+    def parameters(self) -> dict:
+        """Get `dict` of parameters."""
+
+    @abstractmethod
+    def update_parameters(self, new_parameters: dict):
+        """Update the collection of parameters."""
+
+
+class Estimator(ABC):
+    """Estimator for discrepancy model and data."""
+
+    @abstractmethod
+    def __call__(self) -> float:
+        """Evaluate discrepancy."""
+
+    @abstractmethod
+    def gradient(self) -> list:
+        """Compute the gradient of the data set."""
+
+    @property
+    @abstractmethod
+    def parameters(self) -> dict:
+        """Get `dict` of parameters."""
+
+    @abstractmethod
+    def update_parameters(self, new_parameters: dict):
+        """Update the collection of parameters."""
+
+
+class Kinematics(ABC):
+    """Abstract interface for computation of kinematic variables."""
+
+    @abstractmethod
+    def convert(self, events: dict) -> dict:
+        """Convert a set of momentum tuples (events) to kinematic variables."""
+
+    @abstractmethod
+    def is_within_phase_space(self, events: dict) -> list:
+        """Check which events lie within phase space."""
+
+    @property
+    @abstractmethod
+    def phase_space_volume(self) -> float:
+        """Compute volume of the phase space."""
+
+
+class Optimizer(ABC):
+    """Optimize a fit model to a data set."""
+
+    @abstractmethod
+    def optimize(self, estimator: Estimator, initial_parameters: dict) -> dict:
+        """Execute optimization."""
+
+
+class UniformRealNumberGenerator(ABC):
+    """Abstract class for generating uniform real numbers."""
+
+    @abstractmethod
+    def __call__(self, size: int,
+                 min_value=0.0, max_value=1.0) -> float or list:
```

Comment:
Currently, the types are only as hints to the user and for the API (https://github.com/ComPWA/tensorwaves/pull/27) — there are no checks on these. For better typing, it is better to use the [`typing`](https://docs.python.org/3/library/typing.html) module (from Python 3.5, so should be fine) along with [`mypy`](http://mypy-lang.org/). This can also be added as a test to Travis CI.

Reply:
https://stackoverflow.com/questions/33945261/how-to-specify-multiple-return-types-using-type-hints

Btw you are right, this `or` is not valid. I guess its fine we add this in PR #33

Reply:
https://github.com/ComPWA/tensorwaves/pull/30/commits/541a23598565c09ff105474d92cab4910bf4a9e2